### ISSUE TITLE
Get full objects in GetList

### ISSUE THREAD

**Issue by megakoresh** [PaloAltoNetworks/pango#18]
### Is your feature request related to a problem?

When you need to e.g. export firewall configuration to some other format, to get properties of objects you have to first call `GetList` and then iterate through it, calling `Get` on every item. This is slow and extremely inefficient.

### Describe the solution you'd like

`GetList` should obtain the whole section of the underlying XML in one request and parse it to `Entry` objects, instead of returning a slice of strings, representing object names.

### Describe alternatives you've considered

Alternative has been described above in the problem. Another would have been to write own method in consuming code, but since parse methods and schemas aren't exported, that's not possible.

### Additional context

N/A

**Comment by shinmog** [comment]
@megakoresh 

If your intent is not to perform some operation on every security rule in a rulebase, but rather to export config, I think it might be better to just do `fw.Get(....)` at the security rule xpath instead..?

Could update the various xpath functions to be `Xpath` instead so that the xpaths are accessible to the user to help facilitate this.

Another option could be to use goroutines, and just pass in client object to them.

What do you think?

**Comment by megakoresh** [comment]
> @megakoresh
> 
> If your intent is not to perform some operation on every security rule in a rulebase, but rather to export config, I think it might be better to just do `fw.Get(....)` at the security rule xpath instead..?
> 
> Could update the various xpath functions to be `Xpath` instead so that the xpaths are accessible to the user to help facilitate this.
> 
> Another option could be to use goroutines, and just pass in client object to them.
> 
> What do you think?

I need the schema of firewall objects, as, I imagine, would others with similar use-case, as it helps avoid typos and makes the code less verbose and error-prone. If I simply get the XML I would have to parse it in my code, which would result in plenty of boileplate that already exists in this library. Seems like a waste. And anyway - for the firewall the list of names is not that much different than list of full objects, so I don't really understand why those methods work the way they do in the first place.

I am not sure what you meant by updating functions to be Xpath, but hey, as long as we get the parsed Entry objects so we can utilize the schema in the library and don't have to replicate it's code.

By goroutines I imagine you mean parallelizing requests? That's my current solutions, but it doesn't solve the underlying problem - making a ton of unnecessary requests to the firewall. Not to mention the extra code verbosity.

**Comment by shinmog** [comment]
More questions :)

> I need the schema of firewall objects....

When you say this, do you mean you need the normalized `pango` objects?

Is it just security rules that you're after, or something more?

**Comment by megakoresh** [comment]
> More questions :)
> 
> > I need the schema of firewall objects....
> 
> When you say this, do you mean you need the normalized `pango` objects?
> 
> Is it just security rules that you're after, or something more?

The whole config. I need all of it - basically lists of `addr.Entry`, `router.Entry`, etc. But I think this is getting a bit side-tracked. Can we address the question why does the `GetList` function not work this way in the first place? 

I have worked with PanOS apis for almost two years now and they don't seem to care how large an xml you request. It may even be faster, since it doesn't need to aggregate the names. So why not fetch the whole populated list and parse each item?

**Comment by shinmog** [comment]
So, each namespace's `GetList()` invokes `EntryListUsing()`, which does not actually return a huge swath of XML.  Both `EntryListUsing()` and `MemberListUsing()` append some extra stuff onto the end of the xpath specifically to limit the amount of XML that is returned.  So it's not that pango is not doing anything with it; that data is not returned in the XML to begin with.

You're right, I did not expect people to be using pango directly (instead of Terraform), so it is very cool that this is happening :)  I'm just asking questions because I'm trying to find out if there's a way to help you address your need now - without having to wait for me to add new plumbing / functions into pango.

**Comment by megakoresh** [comment]
> So it's not that pango is not doing anything with it; that data is not returned in the XML to begin with.

Yes I understand that, that was precisely my point - modifying the query to return a list of names might actually even ever so slightly put more load on the PanOS API, because the server would have to aggregate the names from the XML on server side, as oppose to simply returning a chunk of it's xml configuration without processing anything.

It's better for client that doesn't care about data - only names (identifiers) and its RAM consumption (since the full xml can be quite big). The firewall does not care for this optimization.

I have had a little time today to dig around the code. It appears the functions you mentioned are part of an interface and correct me if I am wrong, but I think [this](https://github.com/PaloAltoNetworks/pango/blob/ba6c836f559d73632b8ab9b00f9babd7735ff3e2/client.go#L193) is the line responsible for it returning only names.

So the solution would be to change the call signature of this function such that it would allow to conditionally do this modification. Then, because, if I understand correctly, if the `ans` parameter to either of these is `nil`, it will return raw xml string, the object's `GetList` functions can parse the resulting xml using the return struct appropriate for object type? Right? Or what am I missing?

If this is correct then I have enough to try creating a PR. Let me know what you think.

And no, I don't see a solution to do this without replicating library code. The closest is to call `EntryListUsing()` directly, which would require me to replicate the `xpath()` function of the object type for the operation, since it's not exported.

**Comment by megakoresh** [comment]
> If this is correct then I have enough to try creating a PR. Let me know what you think.

I have spent almost whole day yesterday trying to figure this out. The library has this `container_v1/2` objects abstracting differences between versions, and these make things quite a lot more complicated. For now I seem to be stuck.

**Comment by megakoresh** [comment]
@shinmog Alright I think I have some kind of solution:

We can achieve this without breaking backwards compatibility in the following way:

Introduce a new wrapper for Entry array:

```golang
type EntryContainerMulti struct {
	Entries []Entry `xml:"response>result>address>entry"`
}
```

Implement a simple custom unmarshaller: 

```golang
func (e *EntryContainerMulti) UnmarshalXML(d *xml.Decoder, start xml.StartElement) error {
	for {
		token, err := d.Token()
		if err != nil {
			return err
		}
		switch childToken := token.(type) {
		case xml.StartElement:
			switch childToken.Name.Local {
			case "entry":
				entry := entry_v2{}
				err = d.DecodeElement(&entry, &childToken)
				if err != nil {
					return err
				}
				containerized := container_v2{entry}
				e.Entries = append(e.Entries, containerized.Normalize())
			}
		case xml.EndElement:
			if childToken == start.End() {
				return nil
			}
		}
	}
}
```

Add function to object:

```golang
func (c *FwAddr) GetFullList(vsys string) (EntryContainerMulti, error) {
	c.con.LogQuery("(get) list of address objects")
	path := c.xpath(vsys, nil)
	data, err := c.con.Get(path[:len(path)-1], nil, nil)
	if err != nil {
		return EntryContainerMulti{}, err
	}
	ans := EntryContainerMulti{}
	err = xml.Unmarshal(data, &ans)
	if err != nil {
		return EntryContainerMulti{}, err
	}
	return ans, nil
}
```

Several notes about this:

1. In case of AddressObject the multi-versioned containers are unnecessary because `entry_v2` is just a superset of `entry_v1`. However if you have actually different logic somewhere based on version, then because getting correct object relies on the client, unavailable to Entry structs, using this approach can get quite ugly, e.g. - may have to try one version, then next if one fails.
2. Raw XML is returned here, I don't see any other way. So because of of that your generic error handling logic in `EntryListUsing()` and such will not apply
3. Most likely this is not necessary to apply to member lists, because PanOS does not expand those anyway - the API only returns names.

Should I proceed with this approach to the PR? It's quite a lot of work to add this to every object type so I want to be sure first that there aren't any obvious issues.

**Comment by shinmog** [comment]
> Also, unrelated to this, but why do you have both get and show operations everywhere, even though they appear to be functionally identical?

`get` returns candidate config, `show` returns running config.

I'm mobile right now, will respond to the rest soon.

Just as a side note, I really appreciate the effort you've put in so far.

**Comment by megakoresh** [comment]
So uh... You still mobile? You in a worldwide cruise or something?

**Comment by megakoresh** [comment]
@shinmog Since this issue is clearly ignored (there have been activity on the project since you left your last comment), and I have no more chances to contribute after tomorrow, I went ahead and implemented suggested changes on some of the objects that have the highest impact, despite not getting your confirmation on whether these changes would be welcome. I still hope you can at least merge the PR, especially since it's completely additive.

PS: Telling someone you "appreciate" their effort and then proceeding to ignore them, feels somewhat insulting, not gonna lie.

**Comment by shinmog** [comment]
@megakoresh 

I apologize for the silence on this.  I was pulled into something internally, and #17 had been waiting longer for a resolution; I'll try and be better with this.

So, I've not played with custom unmarshaling at all.  After our initial discussions, my thoughts about how to address is close to what your PR is doing.  Please give me some time to play with your PR and custom unmarshaling.

**Comment by shinmog** [comment]
Marking this as resolved by #21 

Let's just use individual issues to track adding this to other object types.
